In [1]:
import numpy as np

In [2]:
L = np.random.random(100)
L

array([0.34998105, 0.34981269, 0.26730562, 0.79347255, 0.70481787,
       0.55716154, 0.7576974 , 0.68492627, 0.13081856, 0.60861583,
       0.37263963, 0.71855272, 0.01362153, 0.51883949, 0.68749815,
       0.97389927, 0.35517296, 0.33120083, 0.85260114, 0.33580259,
       0.7799449 , 0.51252032, 0.67571444, 0.48006514, 0.63762631,
       0.11051852, 0.83112968, 0.75906246, 0.58214204, 0.91780688,
       0.76856486, 0.59930585, 0.10469049, 0.76316496, 0.00621331,
       0.86944664, 0.79018372, 0.45139803, 0.89218008, 0.60052572,
       0.61415202, 0.51601979, 0.0046806 , 0.40931056, 0.03464208,
       0.41311489, 0.03254158, 0.08549612, 0.45884525, 0.07057827,
       0.2318389 , 0.95511647, 0.48258512, 0.90463979, 0.20582398,
       0.9640413 , 0.361338  , 0.37798536, 0.99351975, 0.47195663,
       0.08281201, 0.83098528, 0.29747486, 0.51476717, 0.74829002,
       0.77894684, 0.68175681, 0.17175341, 0.11550538, 0.59369158,
       0.74123786, 0.92440467, 0.41729168, 0.43054591, 0.86023

#### 01 sum运算

In [3]:
# 在python中有一个sum运算
sum(L)

52.07654488028989

In [4]:
# nparray中也有一个sum运算
np.sum(L)

52.07654488028989

下面比较一下两个的性能差异

In [5]:
big_array = np.random.rand(1000000)
%timeit sum(big_array)
%timeit np.sum(big_array)

86.6 ms ± 1.65 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
1.1 ms ± 7.37 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


#### 最小值&最大值

可以看出来，numpy是非常快的

In [6]:
np.min(big_array)

1.0737619157152878e-06

In [7]:
np.max(big_array)

0.9999986233475646

In [8]:
big_array.min()

1.0737619157152878e-06

从上述代码可以看出，两种方式都可以进行聚合操作，但是比较推荐使用np.xx()这种方式，因为可读性更好

接下来看看对于二维数组来说，聚合操作如何

In [9]:
# 从0到16不包括16，使用reshape方法，设置为4行，列数由reshape方法计算
X = np.arange(16).reshape(4, -1)
X

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11],
       [12, 13, 14, 15]])

In [10]:
np.sum(X)

120

以上这个结果就是array从0一直加到15得到了120，可以看出，聚合操作不论你传进来的是一维数组还是二维数组，都是将里面的所有元素进行一个相加的操作。

我们可能会有这样一种需求，我们会想要求二维数组的每一行，每一列的聚合结果，那么，就需要再加上一个参数，axis。

In [11]:
np.sum(X, axis=0)

array([24, 28, 32, 36])

注意下上面的axis参数，表示沿着哪一个维度的方向进行运算，axis为0，说明是从第一个维度的方向进行运算，这里的第一个维度的方向就是行的方向

In [12]:
np.sum(X, axis=1)

array([ 6, 22, 38, 54])

#### 02 prod聚合操作，将数组中所有的元素相乘

In [13]:
np.prod(X)

0

In [14]:
np.prod(X + 1)

2004189184

#### 03 平均值和中位数

In [15]:
np.mean(X)

7.5

In [16]:
np.median(X)

7.5

#### 04 百分位数

In [17]:
# 参数q表示百分之多少
np.percentile(X, q=50)

7.5

In [19]:
for percent in [0, 25, 50, 75, 100]:
    print(np.percentile(X, q=percent))

0.0
3.75
7.5
11.25
15.0


#### 05 方差

In [20]:
np.var(X)

21.25

#### 06 标准差

In [21]:
np.std(X)

4.6097722286464435